In [1]:
import numpy as np
import math

# Problem statement
*Note*: this problem is an extension of the problem from Round 1.

The goldfish are back with more `SCUBA_GEAR`. Each of the goldfish will have new reserve prices, but they still follow the same distribution as in Round 1.

Your trade options are similar to before. You’ll have two chances to offer a good price. Each one of the goldfish will accept the lowest bid that is over their reserve price. But this time, for your second bid, they also take into account the average of the second bids by other traders in the archipelago. They’ll trade with you when your offer is above the average of all second bids. But if you end up under the average, the probability of a deal decreases rapidly.

To simulate this probability, the PNL obtained from trading with a fish for which your second bid is under the average of all second bids will be scaled by a factor *p*:

$$
p = (1000 – \text{average bid}) / (1000 – \text{your bid})
$$

What can you learn from the available data and how will you anticipate on this new dynamic?

# Solution


Let $\bar p$ denote the (ex post) average of second (i.e., highest) bids over all teams in Round 4.

With the notation from Round 1, the surrogate objective function (expected profit for a single fish) is  
$$(p_l, p_h)\mapsto (1000-p_l)\int_{900}^{p_l} (\alpha r + \beta) dr + (1000-p_h)\int_{p_l}^{p_h} (\alpha r + \beta) dr 
\Big(1_{\bar p \leq p_h} + 1_{\bar p > p_h} \frac{1000-\bar p}{1000-p_h}  \Big).$$

In [2]:
def solve(p_avg):
    """Given the average of second bids, find the optimal low and high bids.

    Parameters
    ----------
    p_avg : float
        Average value of second bids.
    
    Returns
    -------
    argmax : list of tuple
        Maximizers and maximal profits.
    """
    val_max = float('-inf')
    argmax = []
    for l in range(900, 1000):
        for h in range(l, 1000):
            temp = (1000 - l) * (-(9/50) * (-900 + l) + (-405000 + l**2/2)/5000)
            temp2 = (1000 - h) * (-(9/50) * (h - l) + (h**2/2 - l**2/2)/5000)
            val = temp + temp2 * (1 if p_avg <= h else (1000-p_avg)/(1000-h))
            if math.isclose(val, val_max):
                argmax.append((l, h, val))
            if val > val_max:
                val_max = val
                argmax = [(l, h, val)]
    return argmax

In [3]:
for p_avg in np.arange(977,1000,0.5):
    print("p_avg:", p_avg, "  Maximizers:", solve(p_avg))

p_avg: 977.0   Maximizers: [(952, 978, 20.41520000000002)]
p_avg: 977.5   Maximizers: [(952, 978, 20.41520000000002)]
p_avg: 978.0   Maximizers: [(952, 978, 20.41520000000002)]
p_avg: 978.5   Maximizers: [(953, 979, 20.409500000000047)]
p_avg: 979.0   Maximizers: [(953, 979, 20.409500000000047)]
p_avg: 979.5   Maximizers: [(953, 980, 20.384300000000053)]
p_avg: 980.0   Maximizers: [(953, 980, 20.384300000000053)]
p_avg: 980.5   Maximizers: [(954, 981, 20.339100000000045)]
p_avg: 981.0   Maximizers: [(954, 981, 20.339100000000045)]
p_avg: 981.5   Maximizers: [(955, 982, 20.270700000000016)]
p_avg: 982.0   Maximizers: [(955, 982, 20.270700000000016)]
p_avg: 982.5   Maximizers: [(955, 983, 20.18130000000001)]
p_avg: 983.0   Maximizers: [(955, 983, 20.18130000000001)]
p_avg: 983.5   Maximizers: [(956, 984, 20.07039999999997)]
p_avg: 984.0   Maximizers: [(956, 984, 20.07039999999997)]
p_avg: 984.5   Maximizers: [(957, 985, 19.934699999999975)]
p_avg: 985.0   Maximizers: [(957, 985, 19.93469

As long as $\bar p \leq 978$, optimal bids are the same as in Round 1.  
When $\bar p > 978$, the maximal profit decreases and the maximizers change as well.

If all the teams seeked to maximize their own profit, they would all play $(952, 978)$.

# Results

As I expected $\bar p$ to be slightly above $978$, I posited that $\bar p = 979$ and played $(953, 979)$.

<img src="https://i.imgur.com/gTAU4NX.png" width="1200" />

The ex post value of $\bar p$ was $980$.